syntax = "proto3";

package omnipricex.scraper;

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";
import "google/protobuf/struct.proto";

// Scraper Service Definition
service ScraperService {
    // Start a new scrape job
    rpc StartScrapeJob(StartScrapeJobRequest) returns (ScrapeJobStatusResponse);
    
    // Get the status of a scrape job
    rpc GetScrapeJobStatus(GetScrapeJobStatusRequest) returns (ScrapeJobStatusResponse);
    
    // Cancel a running scrape job
    rpc CancelScrapeJob(CancelScrapeJobRequest) returns (ScrapeJobStatusResponse);
    
    // List scrape jobs
    rpc ListScrapeJobs(ListScrapeJobsRequest) returns (ListScrapeJobsResponse);
    
    // Get scraped data for a job
    rpc GetScrapedData(GetScrapedDataRequest) returns (GetScrapedDataResponse);
    
    // Manually trigger a scrape for specific URLs
    rpc ScrapeUrls(ScrapeUrlsRequest) returns (ScrapeJobStatusResponse);
}

// Messages
message ScrapeJob {
    string id = 1;
    string job_type = 2;
    string name = 3;
    string status = 4;
    double progress = 5;
    int32 items_total = 6;
    int32 items_scraped = 7;
    int32 items_failed = 8;
    google.protobuf.Timestamp created_at = 9;
    google.protobuf.Timestamp started_at = 10;
    google.protobuf.Timestamp completed_at = 11;
    string celery_task_id = 12;
    string competitor_id = 13;
    string competitor_name = 14;
}

message ScrapedDataItem {
    string id = 1;
    string scrape_job_id = 2;
    string competitor_id = 3;
    string source_url = 4;
    string product_id = 5;
    string product_sku = 6;
    string title = 7;
    double price = 8;
    string currency = 9;
    bool in_stock = 10;
    google.protobuf.Struct extracted_data = 11;
    google.protobuf.Timestamp scraped_at = 12;
    string error_message = 13;
}

// Request/Response messages
message StartScrapeJobRequest {
    string job_type = 1;
    string name = 2;
    repeated string target_urls = 3;
    string competitor_id = 4;
    repeated string product_ids = 5;
    google.protobuf.Struct scraper_config = 6;
    int32 priority = 7;
}

message GetScrapeJobStatusRequest {
    string job_id = 1;
}

message CancelScrapeJobRequest {
    string job_id = 1;
}

message ListScrapeJobsRequest {
    int32 page = 1;
    int32 limit = 2;
    optional string status = 3;
    optional string job_type = 4;
    optional string competitor_id = 5;
}

message ListScrapeJobsResponse {
    repeated ScrapeJob jobs = 1;
    int32 total = 2;
    int32 page = 3;
    int32 limit = 4;
    bool has_next = 5;
}

message GetScrapedDataRequest {
    string job_id = 1;
    int32 page = 2;
    int32 limit = 3;
    optional bool only_successful = 4;
}

message GetScrapedDataResponse {
    repeated ScrapedDataItem data = 1;
    int32 total = 2;
    int32 page = 3;
    int32 limit = 4;
    bool has_next = 5;
}

message ScrapeUrlsRequest {
    repeated string urls = 1;
    string competitor_id = 2;
    int32 priority = 3;
    google.protobuf.Struct config = 4;
}

message ScrapeJobStatusResponse {
    bool success = 1;
    string message = 2;
    ScrapeJob job = 3;
}
